### ISSUE TITLE
v1.12.0-rc.2 generates incompatible CDI spec

### ISSUE THREAD

**Issue by mpolden** [NVIDIA/nvidia-container-toolkit#38]
The CDI spec generated by `nvidia-ctk cdi generate` appears incompatible with Podman 4.2.0 (latest version available in RHEL 8). Podman 4.2.0 is compiled with container-device-interface v0.4.0 (https://github.com/containers/podman/blob/7fe5a419cfd2880df2028ad3d7fd9378a88a04f4/go.mod#L11).

When attempting to use the generated spec, Podman logs the following message:
```
podman run --rm -it --log-level=debug --device nvidia.com/gpu=all docker.io/nvidia/cuda:11.6.2-base-ubuntu20.04 nvidia-smi
...
DEBU[0000] The following errors were triggered when creating the CDI registry: map[/etc/cdi/nvidia.yaml:[failed to load CDI Spec: failed to parse CDI Spec "/etc/cdi/nvidia.yaml": failed to unmarshal CDI Spec: error unmarshaling JSON: while decoding JSON: json: unknown field "hostPath"]]
...
```

This seems to be caused by the presence of the `hostPath` field on device nodes (https://github.com/container-orchestrated-devices/container-device-interface/blob/baded629aa087874924d54af2fb49a87b09f3329/specs-go/config.go#L34), as this field is not present in v0.4.0. Removing this field from the generated file allows the container to start.

The generated spec claims to be version 0.4.0 so it should presumably be compatible? Contents of `/etc/cdi/nvidia.yaml`: 

```yaml
---
cdiVersion: 0.4.0
containerEdits:
  deviceNodes:
  - path: /dev/nvidia-modeset
  - path: /dev/nvidia-uvm-tools
  - path: /dev/nvidia-uvm
  - path: /dev/nvidiactl
  hooks:
  - args:
    - nvidia-ctk
    - hook
    - update-ldcache
    - --folder
    - /usr/lib64
    hookName: createContainer
    path: /bin/nvidia-ctk
  - args:
    - nvidia-ctk
    - hook
    - chmod
    - --mode
    - "755"
    - --path
    - /dev/dri
    hookName: createContainer
    path: /bin/nvidia-ctk
  mounts:
  - containerPath: /usr/lib64/libnvoptix.so.520.61.05
    hostPath: /usr/lib64/libnvoptix.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-wayland-client.so.520.61.05
    hostPath: /usr/lib64/libnvidia-wayland-client.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-vulkan-producer.so.520.61.05
    hostPath: /usr/lib64/libnvidia-vulkan-producer.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-tls.so.520.61.05
    hostPath: /usr/lib64/libnvidia-tls.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-rtcore.so.520.61.05
    hostPath: /usr/lib64/libnvidia-rtcore.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-ptxjitcompiler.so.520.61.05
    hostPath: /usr/lib64/libnvidia-ptxjitcompiler.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-opticalflow.so.520.61.05
    hostPath: /usr/lib64/libnvidia-opticalflow.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-opencl.so.520.61.05
    hostPath: /usr/lib64/libnvidia-opencl.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-nvvm.so.520.61.05
    hostPath: /usr/lib64/libnvidia-nvvm.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-ngx.so.520.61.05
    hostPath: /usr/lib64/libnvidia-ngx.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-ml.so.520.61.05
    hostPath: /usr/lib64/libnvidia-ml.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-gtk3.so.520.61.05
    hostPath: /usr/lib64/libnvidia-gtk3.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-glvkspirv.so.520.61.05
    hostPath: /usr/lib64/libnvidia-glvkspirv.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-glsi.so.520.61.05
    hostPath: /usr/lib64/libnvidia-glsi.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-glcore.so.520.61.05
    hostPath: /usr/lib64/libnvidia-glcore.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-fbc.so.520.61.05
    hostPath: /usr/lib64/libnvidia-fbc.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-encode.so.520.61.05
    hostPath: /usr/lib64/libnvidia-encode.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-eglcore.so.520.61.05
    hostPath: /usr/lib64/libnvidia-eglcore.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-compiler.so.520.61.05
    hostPath: /usr/lib64/libnvidia-compiler.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-cfg.so.520.61.05
    hostPath: /usr/lib64/libnvidia-cfg.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-allocator.so.520.61.05
    hostPath: /usr/lib64/libnvidia-allocator.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvcuvid.so.520.61.05
    hostPath: /usr/lib64/libnvcuvid.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libcudadebugger.so.520.61.05
    hostPath: /usr/lib64/libcudadebugger.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libcuda.so.520.61.05
    hostPath: /usr/lib64/libcuda.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libGLX_nvidia.so.520.61.05
    hostPath: /usr/lib64/libGLX_nvidia.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libGLESv2_nvidia.so.520.61.05
    hostPath: /usr/lib64/libGLESv2_nvidia.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libGLESv1_CM_nvidia.so.520.61.05
    hostPath: /usr/lib64/libGLESv1_CM_nvidia.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libEGL_nvidia.so.520.61.05
    hostPath: /usr/lib64/libEGL_nvidia.so.520.61.05
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /bin/nvidia-smi
    hostPath: /bin/nvidia-smi
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /bin/nvidia-debugdump
    hostPath: /bin/nvidia-debugdump
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /bin/nvidia-persistenced
    hostPath: /bin/nvidia-persistenced
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /bin/nvidia-cuda-mps-control
    hostPath: /bin/nvidia-cuda-mps-control
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /bin/nvidia-cuda-mps-server
    hostPath: /bin/nvidia-cuda-mps-server
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/lib64/libnvidia-egl-gbm.so.1.1.0
    hostPath: /usr/lib64/libnvidia-egl-gbm.so.1.1.0
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/share/glvnd/egl_vendor.d/10_nvidia.json
    hostPath: /usr/share/glvnd/egl_vendor.d/10_nvidia.json
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
  - containerPath: /usr/share/egl/egl_external_platform.d/15_nvidia_gbm.json
    hostPath: /usr/share/egl/egl_external_platform.d/15_nvidia_gbm.json
    options:
    - ro
    - nosuid
    - nodev
    - bind
    type: bind
devices:
- containerEdits:
    deviceNodes:
    - path: /dev/nvidia0
    - hostPath: /dev/dri/card0
      path: /dev/dri/card0
    - hostPath: /dev/dri/renderD128
      path: /dev/dri/renderD128
    hooks:
    - args:
      - /bin/nvidia-ctk
      - hook
      - create-symlinks
      - --link
      - ../card0::/dev/dri/by-path/pci-0000:00:1e.0-card
      - --link
      - ../renderD128::/dev/dri/by-path/pci-0000:00:1e.0-render
      hookName: createContainer
      path: /bin/nvidia-ctk
  name: gpu0
- containerEdits:
    deviceNodes:
    - path: /dev/nvidia0
    - hostPath: /dev/dri/card0
      path: /dev/dri/card0
    - hostPath: /dev/dri/renderD128
      path: /dev/dri/renderD128
  name: all
kind: nvidia.com/gpu
```

**Comment by elezar** [comment]
Yes, we are aware of this issue and will release a fix with an `rc.3` by early next week. Note since the `hostPath` is the same as `Path` for the `deviceNodes` in the edits, these can be removed from the spec and have it work.

**Comment by elezar** [comment]
@mpolden we have released `v1.12.0` which should generate a minimal spec version.

Note that in this case if would be required to run:
```
nvidia-ctk cdi generate --device-name-strategy=type-index
```
or
```
nvidia-ctk cdi generate --device-name-strategy=uuid
``` 
to generate a spec where device names start with an alphabetic character as required by the `v0.4.0` CDI specification.

Please reopen this issue if this does not address your problems.
